### PR TITLE
Add SameAsIncompatibleTypesAnalyzer

### LINKED ISSUE #154: Warning when SameAs compares expressions of incompatible types
```cs
class Foo { }
class Bar { }

var foo = new Foo();
var bar = new Bar();

// Should warn because there is no way that the same instance can be of type Foo and type Bar
Assert.That(foo, Is.SameAs(bar));
```

I think something like bidirectional IsAssignableFrom semantics are desired but I'm not quite sure.

### PR DESCRIPTION
**Pull request by Dreamescaper**
Fixes #154

### CODE REVIEW

**Comment by Dreamescaper** [comment]
Forgot about delegate and Tasks actual types handling. Will add a bit later.

**Comment by Dreamescaper** [comment]
Updated PR.
Added support for any awaitable, therefore will submit a PR for https://github.com/nunit/nunit.analyzers/issues/123 after this one merged (decided not to add it to this PR, even though not much changes there).

**Comment by mikkelbu** [comment]
Thanks for another great contribution @Dreamescaper. I'll probably have time to review this one of the following days. I just merged the other PR, so there are some conflicts - minor as it is due to the same identifier used in both PRs.

**Comment by Dreamescaper** [comment]
Sure, no rush!
Will resolve conflicts later.

File: src/nunit.analyzers.tests/DocumentationTests.cs [hunk 1]
```diff
@@ -176,9 +176,9 @@ int TableLength()
                 var length = 0;
                 while (startIndex + length < doc.Length)
                 {
-                    if (doc[startIndex + length] == '\n' &&
-                        doc.Length > startIndex + length &&
-                        doc[startIndex + length +1] != '|')
+                    if (doc.Length > startIndex + length &&
```
Comment (Dreamescaper) [inline src/nunit.analyzers.tests/DocumentationTests.cs]: It crashed with index out of bound before, if there was no empty line at the end in 'index.md'

Comment (mikkelbu) [reply src/nunit.analyzers.tests/DocumentationTests.cs]: I can remember I added an empty line at some point as a quick solution, but I never came back to correct the code. So thanks for correcting this :)

File: src/nunit.analyzers.tests/SameAsIncompatibleTypes/SameAsIncompatibleTypesAnalyzerTests.cs [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+    public class SameAsIncompatibleTypesAnalyzerTests
+    {
+        private static readonly DiagnosticAnalyzer analyzer = new SameAsIncompatibleTypesAnalyzer();
+        private static readonly ExpectedDiagnostic expectedDiagnostic =
+            ExpectedDiagnostic.Create(AnalyzerIdentifiers.SameAsIncompatibleTypes);
+
+        [Test]
+        public void AnalyzeWhenIncompatibleTypesProvided()
+        {
+            var testCode = TestUtility.WrapClassInNamespaceAndAddUsing($@"
```
Comment (mikkelbu) [inline src/nunit.analyzers.tests/SameAsIncompatibleTypes/SameAsIncompatibleTypesAnalyzerTests.cs]: I think you can remove the $-part from all strings in this file and then also avoid the `{{` and `}}`. I can also do this and push an update to your branch, if you don't mind?

Comment (Dreamescaper) [reply src/nunit.analyzers.tests/SameAsIncompatibleTypes/SameAsIncompatibleTypesAnalyzerTests.cs]: Sure I don't mind, thanks!

**Comment by mikkelbu** [comment]
Thanks for the contribution @Dreamescaper

**Pull request merged**